File: server/src/engine-manager.ts
import type { EngineAdapter } from './adapters/types.js';
import type { EngineInfo } from '../../shared/types.js';

/**
 * Registry of engine adapters.
 * The server creates one EngineManager at startup, registers adapters,
 * and looks them up by id when handling API requests.
 */
export class EngineManager {
  private adapters = new Map<string, EngineAdapter>();

  register(adapter: EngineAdapter): void {
    if (this.adapters.has(adapter.id)) {
      throw new Error(`Adapter "${adapter.id}" is already registered`);
    }
    this.adapters.set(adapter.id, adapter);
  }

  get(engineId: string): EngineAdapter {
    const adapter = this.adapters.get(engineId);
    if (!adapter) {
      throw new Error(`Unknown engine "${engineId}"`);
    }
    return adapter;
  }

  list(): EngineInfo[] {
    return [...this.adapters.values()].map((a) => ({
      id: a.id,
      name: a.name,
      type: a.type,
      connected: a.isConnected(),
    }));
  }
}

File: server-py/engine_manager.py
"""Registry of engine adapters."""

from __future__ import annotations

from adapters.base import EngineAdapter
from models import EngineInfo


class EngineManager:
    def __init__(self) -> None:
        self._adapters: dict[str, EngineAdapter] = {}

    def register(self, adapter: EngineAdapter) -> None:
        if adapter.id in self._adapters:
            raise ValueError(f'Adapter "{adapter.id}" is already registered')
        self._adapters[adapter.id] = adapter

    def get(self, engine_id: str) -> EngineAdapter:
        adapter = self._adapters.get(engine_id)
        if adapter is None:
            raise ValueError(f'Unknown engine "{engine_id}"')
        return adapter

    def list(self) -> list[EngineInfo]:
        return [
            EngineInfo(
                id=a.id,
                name=a.name,
                type=a.engine_type,
                connected=a.is_connected(),
            )
            for a in self._adapters.values()
        ]
